[DCJ-503] Gracefully handle DRS resolution failure case for self-hosted snapshots

okotsopoulos opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-503

## Addresses

In production, we are seeing some DRS resolution attempts fail with an NPE:

```
java.lang.NullPointerException: Cannot invoke \"com.google.cloud.storage.Bucket.getLocation()\" because \"bucket\" is null"
```

Reading the corresponding stack trace shows that this is happening for self-hosted GCP snapshots ([logs](https://cloudlogging.app.goo.gl/AiK6u3JBHga1BYFZ8)).

## Summary of changes

- Gracefully handle the case where TDR cannot obtain the GCS bucket for a file object within a self-hosted GCP snapshot. 
-  Provide diagnostic information on the source snapshot, root dataset, and file object in the error details.  (Note that if we've gotten this far in our attempted DRS resolution, we've already verified that the user is authorized to read the snapshot, so they're allowed to get this diagnostics information back.)

An example exception thrown with this change (obtained from unit tests, so may not be fully populated as we've seen in the wild):
```
DrsObjectNotFoundException
GCS bucket from gs://path/to/file.txt not found (diagnostics returned in error details)
[class SnapshotCacheResult {
  snapshot: {
    id: e1c97007-d80a-487d-90da-f019ba6a9ebd
    name: null
    isSelfHosted: true
    globalFileIds: false
    billingProfileId: bd9b1387-2fb3-4aad-959f-c6015cdaab8a
    cloudPlatform: gcp
    googleProjectId: snapshot-google-project
  }
  rootDataset: {
    id: null
    name: null
    datasetProjectId: dataset-google-project
  }
}, bio.terra.service.filedata.FSFile@3cb01485[datasetId=<null>,cloudPath=gs://path/to/file.txt,mimeType=<null>,bucketResourceId=1fc4ea5c-fdc7-4262-ba38-7a8dcfbeb89b,loadTag=<null>]]
```

## Testing Strategy

Expanded unit tests.

<!-- Reminder -->
<!-- Two CODEOWNERS will be automatically assigned to review this pull request. If you otherwise have two reviewers, you do not need to wait for their review. -->

okotsopoulos committed:
[DCJ-503] Gracefully handle DRS resolution failure case for self-hosted snapshots

In production, we are seeing some DRS resolution attempts fail with an NPE:

java.lang.NullPointerException: Cannot invoke \"com.google.cloud.storage.Bucket.getLocation()\" because \"bucket\" is null"

Reading the corresponding stack trace shows that this is happening for self-hosted GCP snapshots.

Gracefully handle the case where TDR cannot obtain the GCS bucket for a file object within a self-hosted GCP snapshot, and provide diagnostic information on the source snapshot, root dataset, and file object in the error details.

okotsopoulos committed:
Add unit tests

okotsopoulos committed:
Expand unit test to confirm authorization verified

rushtong reviewed (approved):
Thank you 👍🏽

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -1069,5 +1074,27 @@ public UUID getSnapshotBillingProfileId() {
       return this.cloudPlatform;
     }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("class SnapshotCacheResult {\n");
+      sb.append("  snapshot: {\n");
+      sb.append("    id: ").append(this.id).append("\n");
+      sb.append("    name: ").append(this.name).append("\n");
+      sb.append("    isSelfHosted: ").append(this.isSelfHosted).append("\n");
+      sb.append("    globalFileIds: ").append(this.globalFileIds).append("\n");
+      sb.append("    billingProfileId: ").append(this.snapshotBillingProfileId).append("\n");
+      sb.append("    cloudPlatform: ").append(this.cloudPlatform).append("\n");
+      sb.append("    googleProjectId: ").append(this.googleProjectId).append("\n");
+      sb.append("  }\n");
+      sb.append("  rootDataset: {\n");
+      sb.append("    id: ").append(this.datasetId).append("\n");
+      sb.append("    name: ").append(this.datasetName).append("\n");
+      sb.append("    datasetProjectId: ").append(this.datasetProjectId).append("\n");
+      sb.append("  }\n");
+      sb.append("}");
+      return sb.toString();
+    }
```
It's longer but I think a text block makes it clearer what the output will look like:
```
    @Override
    public String toString() {
      return """
          class SnapshotCacheResult {
            snapshot: {
              id: %s
              name: %s
              isSelfHosted: %s
              globalFileIds: %s
              billingProfileId: %s
              cloudPlatform: %s
              googleProjectId: %s
            }
            rootDataset: {
              id: %s
              name: %s
              datasetProjectId: %s
            }
          }
          """
          .formatted(
              id,
              name,
              isSelfHosted,
              globalFileIds,
              snapshotBillingProfileId,
              cloudPlatform,
              googleProjectId,
              datasetId,
              datasetName,
              datasetProjectId);
    }
```

okotsopoulos replied:
I agree with you.  I'll resolve this convo as it's no longer needed with a refactor to a record.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -716,11 +716,16 @@ private String retrieveGCPSnapshotRegion(SnapshotCacheResult cachedSnapshot, FSF
-              // ...but bill to the snapshot's project
-              BucketGetOption.userProject(cachedSnapshot.googleProjectId));
+      // ...but bill to the snapshot's project.
+      BucketGetOption userProject = BucketGetOption.userProject(cachedSnapshot.googleProjectId);
```
Looking at the constructor for `SnapshotCacheResult`, it looks possible to have a null `googleProjectId` even if `isSelfHosted` is true. If that happens, I think `BucketGetOption.userProject()` will throw an NPE.

If the code is correct -- when `isSelfHosted` is true, `googleProjectId` is not null -- you can make it explicit in the code by using `requireNonNull()`
```suggestion
      BucketGetOption userProject = BucketGetOption.userProject(Objects.requireNonNull(cachedSnapshot.googleProjectId));
```

okotsopoulos replied:
Good shout -- I took a closer look at this.  The guarantee we can rely on is as follows:

_When a snapshot's backing cloud platform is GCP, `googleProjectId` is not null.  The snapshot's self-hosted status has no bearing on `googleProjectId`'s value._

We have a shaky chain of method definitions to get here: this code snippet lives in a method called `retrieveGCPSnapshotRegion` which should only be called for a GCP-backed snapshot.  I am inclined to expand docstrings and comments to make the chain of guarantee more clear, and save any further refactoring for the future team.  We don't currently use `Objects.requireNonNull` in production code, and it looks like in this case it would be a targeted bandaid in place of a refactor to platform-specific snapshot and/or cached snapshots.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -716,11 +716,16 @@ private String retrieveGCPSnapshotRegion(SnapshotCacheResult cachedSnapshot, FSF
+        throw new DrsObjectNotFoundException(
+            "GCS bucket from %s not found (diagnostics returned in error details)"
+                .formatted(cloudPath),
+            List.of(cachedSnapshot.toString(), fsFile.toString()));
```
Why are some data provided in the initial error message and some are appended at the end? Could all the data be provided in the error message? My understanding was `causes` is used in the case where we want to collect the chained exception causes and roll them up to the top level message.

okotsopoulos replied:
We haven't been perfectly consistent here, but the way that this exception will ultimately be massaged into a response to the user treats any optional supplemental elements as "error details":

https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/resources/api/data-repository-openapi.yaml#L6770-L6777

Because the stringified `cachedSnapshot` and `fsFile` are both large and complex, I believe that embedding them in the top-level error message will make it harder to diagnose the issue.  This is something that could be easily changed in the future if devs and users preferred.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -1069,5 +1074,27 @@ public UUID getSnapshotBillingProfileId() {
     }
+
+    @Override
+    public String toString() {
```
Instead of a custom `toString()`, how about converting this class to a `record`? It looks like all the fields are read-only already.
```
  record SnapshotCacheResult(
      UUID id,
      String name,
      boolean isSelfHosted,
      boolean globalFileIds,
      BillingProfileModel datasetBillingProfileModel,
      UUID snapshotBillingProfileId,
      CloudPlatform cloudPlatform,
      String googleProjectId,
      String datasetProjectId,
      UUID datasetId,
      String datasetName) {

    public SnapshotCacheResult(Snapshot snapshot) {
      this(
          snapshot.getId(),
          snapshot.getName(),
          snapshot.isSelfHosted(),
          snapshot.hasGlobalFileIds(),
          snapshot.getSourceDataset().getDatasetSummary().getDefaultBillingProfile(),
          snapshot.getProfileId(),
          snapshot.getCloudPlatform(),
          Optional.ofNullable(snapshot.getProjectResource())
              .map(GoogleProjectResource::getGoogleProjectId)
              .orElse(null),
          Optional.ofNullable(snapshot.getSourceDataset().getProjectResource())
              .map(GoogleProjectResource::getGoogleProjectId)
              .orElse(null),
          snapshot.getSourceDataset().getId(),
          snapshot.getSourceDataset().getName());
    }
```

okotsopoulos replied:
Great idea!  Working on that now.

pshapiro4broad reviewed (approved):
Looks OK, but using a `record` may be a better way to ensure that all fields in a class are logged.

fboulnois reviewed (approved):
I agree with Phil's comment on string formatting. Otherwise looks good 👍

okotsopoulos committed:
Merge branch 'develop' into okotsopo-DCJ-503-drs-selfhosted-diagnostics

okotsopoulos committed:
PR feedback - refactor SnapshotCacheResult to record

Eliminating the need for a custom toString definition.
Expanded unit tests to verify diagnostic information returned to caller.

okotsopoulos committed:
Expand docstrings, comments
